### PR TITLE
Prevent BOM Versions Overlapping

### PR DESCRIPTION
**Pull request by riadmen**
#17000

### CODE REVIEW

File: backend/de.metas.manufacturing/src/main/java/org/eevolution/model/validator/PP_Product_BOM.java [hunk 1]
```diff
@@ -109,4 +121,69 @@ public void validateBOMAttributes(final I_PP_Product_BOM productBom)
...
+				{
+					throw new AdempiereException(MSG_BOM_VERSIONS_OVERLAPPING,
+												 productBom.getName())
+							.markAsUserValidationError();
+				}
+			}
+		}
+	}
+
+	private boolean isOverlapping(final Timestamp newFrom, final Timestamp newTo, final Timestamp existingFrom, final Timestamp existingTo)
```
Comment (teosarca) [inline backend/de.metas.manufacturing/src/main/java/org/eevolution/model/validator/PP_Product_BOM.java:165]: pls
1. move it to TimeUtil and make it static
2. annotate params with `@Nullable` or `@NonNull`
3. provide unit tests including edge cases to understand if the intervals are [closed,open) or [closed,closed]. I would suggest to go with [closed,open) interval

File: backend/de.metas.manufacturing/src/main/java/org/eevolution/model/validator/PP_Product_BOM.java [hunk 2]
```diff
@@ -109,4 +121,69 @@ public void validateBOMAttributes(final I_PP_Product_BOM productBom)
...
+	public void preventBOMVersionsOverlapping(final I_PP_Product_BOM productBom)
+	{
+		if (productBom.getValidTo() != null && productBom.getValidTo().before(productBom.getValidFrom()))
+		{
+			throw new AdempiereException(MSG_VALID_TO_BEFORE_VALID_FROM);
+		}
+
+		final ProductBOMVersionsId bomVersionsId = ProductBOMVersionsId.ofRepoId(productBom.getPP_Product_BOMVersions_ID());
+		final BOMType bomType = BOMType.ofNullableCode(productBom.getBOMType());
+		final Optional<I_PP_Product_BOM> latestBOMVersions = bomDAO.getLatestBOMRecordByVersionAndType(bomVersionsId, ImmutableSet.of(bomType));
```
Comment (teosarca) [inline backend/de.metas.manufacturing/src/main/java/org/eevolution/model/validator/PP_Product_BOM.java:136]: why not retrieve and check all BOMs of bomVersionsId and bomType?
why check only the latest one?

**Review by teosarca — CHANGES_REQUESTED** [review]
check my comments

File: backend/de.metas.manufacturing/src/main/java/org/eevolution/model/validator/PP_Product_BOM.java (hunk 2 above)
Comment (riadmen) [reply backend/de.metas.manufacturing/src/main/java/org/eevolution/model/validator/PP_Product_BOM.java]: code updated

File: backend/de.metas.business/src/main/java/org/eevolution/api/impl/ProductBOMDAO.java [hunk 3]
```diff
@@ -419,6 +419,20 @@ public Optional<I_PP_Product_BOM> getPreviousVersion(final @NonNull I_PP_Product
 				.firstOptional(I_PP_Product_BOM.class);
 	}
 
+	@Override
+	@NonNull
+	public List<I_PP_Product_BOM> retrieveAllBomVersions(final @NonNull I_PP_Product_BOM bomVersion)
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/org/eevolution/api/impl/ProductBOMDAO.java:424]: either u call this method sometihng like getSiblings
or you remove 
> .addNotEqualsFilter(I_PP_Product_BOM.COLUMNNAME_PP_Product_BOM_ID, bomVersion.getPP_Product_BOM_ID())

and u handle this in your caller.

----

Also
> final @NonNull I_PP_Product_BOM bomVersion

pls rename it to `bom` because it's NOT a `bomVersions`

File: backend/de.metas.adempiere.adempiere/base/src/main/java-legacy/org/compiere/util/TimeUtil.java [hunk 4]
```diff
@@ -2110,4 +2110,24 @@ public static long getMillisBetween(@NonNull final Timestamp timestamp1, @NonNul
 	{
 		return timestamp2.getTime() - timestamp1.getTime();
 	}
+
+	public static boolean isOverlapping(@NonNull final Timestamp newValidFrom, @Nullable final Timestamp newValidTo, @NonNull final Timestamp existingValidFrom, @Nullable final Timestamp existingValidTo)
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/base/src/main/java-legacy/org/compiere/util/TimeUtil.java:2114]: rename those params as start1, end1, start2, end2

File: backend/de.metas.manufacturing/src/main/java/org/eevolution/model/validator/PP_Product_BOM.java [hunk 5]
```diff
@@ -109,4 +120,34 @@ public void validateBOMAttributes(final I_PP_Product_BOM productBom)
...
+	@ModelChange(timings = { ModelValidator.TYPE_BEFORE_NEW, ModelValidator.TYPE_BEFORE_CHANGE },
+			ifColumnsChanged = { I_PP_Product_BOM.COLUMNNAME_ValidFrom, I_PP_Product_BOM.COLUMNNAME_ValidTo })
+	public void preventBOMVersionsOverlapping(final I_PP_Product_BOM productBom)
+	{
+		if (productBom.getValidTo() != null && productBom.getValidTo().before(productBom.getValidFrom()))
+		{
+			throw new AdempiereException(MSG_VALID_TO_BEFORE_VALID_FROM);
+		}
+
+		final List<I_PP_Product_BOM> allBOMVersions = bomDAO.getSiblings(productBom);
```
Comment (teosarca) [inline backend/de.metas.manufacturing/src/main/java/org/eevolution/model/validator/PP_Product_BOM.java:133]: inline this or rename the variable to something like "siblingBOMs"

File: backend/de.metas.manufacturing/src/main/java/org/eevolution/model/validator/PP_Product_BOM.java [hunk 6]
```diff
@@ -109,4 +120,34 @@ public void validateBOMAttributes(final I_PP_Product_BOM productBom)
...
+
+		for (final I_PP_Product_BOM existingBOMVersion : allBOMVersions) {
+
+			if (productBom.getPP_Product_BOM_ID() == existingBOMVersion.getPP_Product_BOM_ID()) {
```
Comment (teosarca) [inline backend/de.metas.manufacturing/src/main/java/org/eevolution/model/validator/PP_Product_BOM.java:137]: IMHO this is no longer needed

File: backend/de.metas.manufacturing/src/main/java/org/eevolution/model/validator/PP_Product_BOM.java [hunk 7]
```diff
@@ -109,4 +120,34 @@ public void validateBOMAttributes(final I_PP_Product_BOM productBom)
...
+				continue;
+			}
+
+			final Timestamp newValidFrom = productBom.getValidFrom();
+			final Timestamp newValidTo = productBom.getValidTo();
+			final Timestamp existingValidFrom = existingBOMVersion.getValidFrom();
+			final Timestamp existingValidTo = existingBOMVersion.getValidTo();
```
Comment (teosarca) [inline backend/de.metas.manufacturing/src/main/java/org/eevolution/model/validator/PP_Product_BOM.java:144]: it's a matter of taste, but IMHO you could inline those variables

File: backend/de.metas.adempiere.adempiere/base/src/main/java-legacy/org/compiere/util/TimeUtil.java (hunk 4 above)
Comment (riadmen) [reply backend/de.metas.adempiere.adempiere/base/src/main/java-legacy/org/compiere/util/TimeUtil.java]: done

File: backend/de.metas.manufacturing/src/main/java/org/eevolution/model/validator/PP_Product_BOM.java (hunk 5 above)
Comment (riadmen) [reply backend/de.metas.manufacturing/src/main/java/org/eevolution/model/validator/PP_Product_BOM.java]: done

File: backend/de.metas.manufacturing/src/main/java/org/eevolution/model/validator/PP_Product_BOM.java (hunk 7 above)
Comment (riadmen) [reply backend/de.metas.manufacturing/src/main/java/org/eevolution/model/validator/PP_Product_BOM.java]: done

File: backend/de.metas.adempiere.adempiere/base/src/main/java-legacy/org/compiere/util/TimeUtil.java [hunk 8]
```diff
@@ -2110,4 +2110,24 @@ public static long getMillisBetween(@NonNull final Timestamp timestamp1, @NonNul
 	{
 		return timestamp2.getTime() - timestamp1.getTime();
 	}
+
+	public static boolean isOverlapping(@NonNull final Timestamp newValidFrom, @Nullable final Timestamp newValidTo, @NonNull final Timestamp start2, @Nullable final Timestamp end2)
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/base/src/main/java-legacy/org/compiere/util/TimeUtil.java:2114]: start1, end1

Comment (teosarca) [reply backend/de.metas.adempiere.adempiere/base/src/main/java-legacy/org/compiere/util/TimeUtil.java:2114]: TODO: junit tests for cases like:
1. 1---1    2--------------2
2. 1---12-------------2
3. 1---2--1-------------2
4. 2---1--1-----------2
5. 2-----------1---2----1
6. 2-----------21----1
7. 2-----------2       1----1

and add tests for null cases!!!!

Comment (teosarca) [reply backend/de.metas.adempiere.adempiere/base/src/main/java-legacy/org/compiere/util/TimeUtil.java:2114]: also assert 
* start1<=end1 , if not null
* start2 <= end2, if not null

File: backend/de.metas.manufacturing/src/main/java/org/eevolution/model/validator/PP_Product_BOM.java [hunk 9]
```diff
@@ -109,4 +120,32 @@ public void validateBOMAttributes(final I_PP_Product_BOM productBom)
 		productPlanningDAO.retrieveProductPlanningForBomVersions(productBOMVersionsId)
 				.forEach(productPlanning -> productBOMService.verifyBOMAssignment(productPlanning, productBom));
 	}
+
+	@ModelChange(timings = { ModelValidator.TYPE_BEFORE_NEW, ModelValidator.TYPE_BEFORE_CHANGE },
+			ifColumnsChanged = { I_PP_Product_BOM.COLUMNNAME_ValidFrom, I_PP_Product_BOM.COLUMNNAME_ValidTo })
+	public void preventBOMVersionsOverlapping(final I_PP_Product_BOM productBom)
+	{
+		if (productBom.getValidTo() != null && productBom.getValidTo().before(productBom.getValidFrom()))
```
Comment (teosarca) [inline backend/de.metas.manufacturing/src/main/java/org/eevolution/model/validator/PP_Product_BOM.java:128]: as it looks like, IMHO you could move the body of this method to `ProductBOMService` as a new method called `assertNoOverlapping` so your interceptor shall be more glue code than actual business logic.

File: backend/de.metas.adempiere.adempiere/base/src/main/java-legacy/org/compiere/util/TimeUtil.java (hunk 8 above)
Comment (riadmen) [reply backend/de.metas.adempiere.adempiere/base/src/main/java-legacy/org/compiere/util/TimeUtil.java]: done

File: backend/de.metas.manufacturing/src/main/sql/postgresql/system/30-de.metas.adempiere.libero/5714350_sys_gh17004_IsVoided_PP_Order.sql [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+INSERT INTO AD_Column (AD_Client_ID,AD_Column_ID,AD_Element_ID,AD_Org_ID,AD_Reference_ID,AD_Table_ID,ColumnName,ColumnSQL,Created,CreatedBy,DDL_NoForeignKey,DefaultValue,EntityType,FacetFilterSeqNo,FieldLength,FilterDefaultValue,FilterOperator,IsActive,IsAdvancedText,IsAllowLogging,IsAlwaysUpdateable,IsAutoApplyValidationRule,IsAutocomplete,IsCalculated,IsDimension,IsDLMPartitionBoundary,IsEncrypted,IsExcludeFromZoomTargets,IsFacetFilter,IsForceIncludeInGeneratedModel,IsGenericZoomKeyColumn,IsGenericZoomOrigin,IsIdentifier,IsKey,IsLazyLoading,IsMandatory,IsParent,IsSelectionColumn,IsShowFilterIncrementButtons,IsShowFilterInline,IsStaleable,IsSyncDatabase,IsTranslated,IsUpdateable,IsUseDocSequence,MaxFacetsToFetch,Name,SelectionColumnSeqNo,SeqNo,Updated,UpdatedBy,Version) VALUES (0,587764,582876,0,20,53027,'IsVoided','(CASE WHEN PP_Order.DocStatus = ''VO'' THEN ''Y'' ELSE ''N'' END)',TO_TIMESTAMP('2023-12-21 12:00:27.852','YYYY-MM-DD HH24:MI:SS.US'),100,'N','N','EE01',0,1,'N','E','Y','N','Y','N','N','N','N','N','N','N','Y','N','N','N','N','N','N','Y','N','N','Y','N','N','N','N','N','N','N',0,'Storniert',0,0,TO_TIMESTAMP('2023-12-21 12:00:27.852','YYYY-MM-DD HH24:MI:SS.US'),100,0)
```
Comment (teosarca) [inline backend/de.metas.manufacturing/src/main/sql/postgresql/system/30-de.metas.adempiere.libero/5714350_sys_gh17004_IsVoided_PP_Order.sql:12]: why, what columns is this? how is handled?

**Review by teosarca — APPROVED** [review]
